[Long-running feature] Bottom-up Content Moderation

tsmacdonald opened the pull request.

daltojohnso committed:
Add Saved Question button header and new, empty details sidebar (#15706)

* Add question button header and details sidebar

* rmv QuestionEntityMenu

* changes to address review comments

daltojohnso committed:
Add QuestionActionButtons to question details sidebar (#15733)

* add/impl QuestionActionButtons component

* add tooltips

daltojohnso committed:
first pass at button to trigger the creation of moderation issues (#15789)

* support button EntityMenuTrigger + a few additional optional item props

* first pass at  ModerationIssueActionMenu

import the ee moderation stuff

rmv unnecessary cx usage

localize

play with keeping more biz logic in lib/moderation

name/status --> type

make the moderation ee plugin feel a little more ergonomic

move rest of moderation logic into ee directory

* rmv lib/moderation and move all logic to ee folder

daltojohnso committed:
Add unwired moderation issue creation sidebar panel (#15793)

* refactor QuestionDetailsSidebar to handle multiple different views

* implement the issue creation form, unwired

* split details panel into separate file

* force textarea to stay width of panel

* fix translations

* add PluginPlaceholder component to avoid branching logic

* make component match filename

* add description to details sidebar + edit action icon (#15840)

daltojohnso committed:
add history button to saved question header (#15837)

tsmacdonald committed:
Bucm/moderation requests (#15838)

* Placeholder commit for the long-running BUCM feature branch

* Basic ModerationRequest model. Needs permissions

* Add moderated_item hydration to ModerationReviews too

* Return moderation reqs/revs with dashboard

* Add moderation requests/reviews to card API

* remove dead code

* cyclic load

* Add Moderation models to cmd.copy

* Adds endpoint to create ModerationReviews

daltojohnso committed:
add Comment and CommentHeader UI components (#15857)

* add Comment and CommentHeader UI components

* add unit tests

tsmacdonald committed:
Let ModerationReviews be created with a status (#15884)

daltojohnso committed:
add class for targeting in unit test

daltojohnso committed:
fix a few cypress tests broken by missing pencil icon

tsmacdonald committed:
Fix E2E test for question duplication

tsmacdonald committed:
Linter fix

daltojohnso committed:
let users create moderation reviews via saved question sidebar (#15885)

* first pass (unfinished) wire-up of mod review

* make it a basic form

* update moderation type names

* wire things up to moderation-reviews endpoint + soft reload of card

* add useAsyncFunction hook

* fix prettier problem

* address review comments

daltojohnso committed:
add open moderation issues panel with mock data (#15918)

* add open issues panel with mock data

* rmv comment

daltojohnso committed:
add timeline with mock data to sidebar (#15944)

* add timeline with mock data to sidebar

* make the sidebar scrollbar show and not mess up panel dimensions

* correctly hide sidebar when CLOSED_NATIVE_EDITOR_SIDEBARS action triggers

* hide border next to final event in timeline

* make each timeline item expand to full width

* show revision history in sidebar timeline

tsmacdonald committed:
Request API (#16063)

* Add MRQ creation; schema updates

* Edit ModerationRequests

* Edit ModerationReviews

* Add Questions import to query builder

daltojohnso committed:
mock moment's fromNow for snapshot tests

mazameli committed:
Styling on BUCM (#16055)

* saved question chevron size

* restyle HistoryButton

* QuestionActionButtons styling

* more heading and subheading styling

* rmv some unused code

* reduce y padding on sidebar moderation buttons

Co-authored-by: Dalton Johnson <daltojohnso@users.noreply.github.com>

mazameli commented:
Two quick things I noticed that are probably already known:
- Moderation actions aren't yet showing up in the Activity history (e.g. "Maz verified this question")
- (Low-prio) We shouldn't be changing the color of a question's title to blue/orange/purple when a moderator badges it. Only the verified/warning/confused icons should be blue/organge/purple.

This screenshot shows both of those things:

![image](https://user-images.githubusercontent.com/2223916/118884861-d822a680-b8ab-11eb-9f60-d3a88884ee94.png)

daltojohnso committed:
Let users create moderation requests and let admins create moderation reviews (on questions) (#16141)

* first pass reviews + requests wire up

* use const struct

* add an isModerator selector to avoid passing props around a lot

* pass moderation_request params to action fn

* update var name since we have multiple uses of word type

* remove outdated comment

* rmv unneeded passing of isAdmin prop

* fix broken cy tests

- rmv assumption of moderation_requests and _reviews existing
- replace cy.icon with a data-testid

* fix arg bug for getting mod action types

* fix history button cy tests + make it a button

* create moderation requests with correct props

* rename fn to reflect that it creates a review OR a request

daltojohnso committed:
Improve saved question sidebar empty states and styling (#16058)

* add clamped description component with empty state + empty action

* basic empty state for open issues panel

* show open issues button when there are 0

* show better text for question creation event in activity timeline

* fix rebase messup wrt question description

* fix saved question header styling

* disable open issues button when there are no open issues

* maz spacing tweaks

* disabled button styling

* update jest snapshots

Co-authored-by: Maz Ameli <maz@metabase.com>

tsmacdonald committed:
Bucm/comment backend (#16136)

* Add migration for comments

* Add Comment model and appropriate hydration

* Comment API

* Return moderation elements when updating a card

* Fix comment fk

tsmacdonald committed:
Close moderation requests when a review is verified (#16219)

* Close moderation requests when a review is verified

* Use real pre-update hooks instead of handrolling them

* Also close requests when a question is misleading or confusing

daltojohnso committed:
add moderation events and footer action buttons to saved question activity timeline (#16195)

* first pass at mod events in timeline

* fix typo

* rmv unused prop

* Show revert button in timeline and make it reload the card

the reloadCard action previously retrieved an "originalCard" from state.
I'm not totally clear _why_, but it seems like the correct course of
action after triggering a revert is to re-GET the card so that it
reflects the revert. It may be a revert from a long time ago and that
would not be stored in client state.

* add footer to moderation request events

* fix localization of mod request statuses in footer

* rmv unused prop

* separate hover styling into own class

* fix linting errs

* rmv unused import

* split timeline footer into two components

* add commenting to moderation requests (#16178)

* add a form for commenting to moderation threads

* add fake service logic

* prevent adding of empty comments

* uncomment out comment action

* wire up commenting to api

* limit comment mod badge to user 1

* autofocus comment input

* replace reloadCard with softReloadCard

* update dismiss action to add description as comment on request

* use autofocus instead of js

* add more copy to bucm (#16243)

mazameli committed:
add document_curled icon (#16277)

daltojohnso committed:
fix err when not EE in question timeline

daltojohnso committed:
fix styling of moderation request buttons in details panel

daltojohnso committed:
Add notification center for BUCM (#16285)

* boilerplate for nav link component + routing

* add some TODOS to NotificationCenterLink

* add notification center header ui

* add ui for requests lists

* show all open mod requests on notifications page

* janky moderation-request GET implt

* jankily wire up to janky moderation-request GET

* pass correct link to request notification card

* only make pill grayscale when passing status

* add janky auto-opening sidebar for navigating from notifications to specific request

* fix bug caused by race condition in updating redux state

* clean up pill styling

* fix getSection fn

* pull out dup'd className and hover styling

* tweak RequestNotificationCard component names

* clean up the request notifications app a little

* pass status to pill in sidebar

* use slug in url from notifications to questions

daltojohnso committed:
add open moderation request action menu (#16509)

* add open moderation request action menu

* ensure comment form is closed when request text form is toggled on

* add request resolved/dismissed events to timeline

* use useAsyncFunction hook

daltojohnso committed:
add mod request + removal actions to moderation action menu (#16529)

* add mod request + removal actions to moderation action menu

* clean up action menu logic a little

* split dismiss into its own action

* a few tweaks from pr review

tsmacdonald committed:
BUCM Notifications (#16543)

* Basic notification endpoint

* Create notifications for comments

* Create notifications for moderation reviews

* Add test

mazameli committed:
change strings (#16548)

tsmacdonald committed:
Bucm/collection perms (#16266)

* Rename 'curate collection' perm to 'edit collection'

* Initial work on moderate permission

* Finish adding moderate permission

* add "moderate" collection permission level

* Let moderate permissions be set

* Use explicit /edit/ permission; use / for moderation

TODO: can-read? is broken

* Make can-read work

* disable duplicate perm-checking in test

* Test fixes

daltojohnso committed:
replace usage of HistoryButton with LastEditInfoLabel
